### PR TITLE
Add debouncing mechanism and before-capture callbacks for screenshots/vh

### PR DESCRIPTION
**Pull request by markushi**
## :scroll: Description

As discussed we want to debounce the capturing of screenshots / view hierarchies. As of now I hard-coded the debouncing time to 2 seconds, thus in worst case capturing is happening every 2 seconds.

On top of that a new `beforeCapture` callback is introduced, to allow even more customization and eventually overrule the debouncing mechanism. This callback can also be used to prevent capturing for non-fatal events.

## :bulb: Motivation and Context

Fixes https://github.com/getsentry/sentry-java/issues/2713 and https://github.com/getsentry/sentry-java/issues/2665

## :pencil: Checklist

- [x] I reviewed the submitted code.
- [x] I added tests to verify the changes.
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java [hunk 1]
```diff
@@ -52,6 +59,16 @@ public ScreenshotEventProcessor(
       return event;
     }
 
+    final boolean shouldDebounce = debouncer.checkForDebounce();
```
Comment (stefanosiano) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:62]: I think everything is fine here
If you could just add a comment to explain the order of the calls it would be even better 😄

File: sentry-android-core/src/test/java/io/sentry/android/core/internal/util/DebouncerTest.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        assertTrue(debouncer.checkForDebounce())
+    }
+
+    @Test
+    fun `Debouncer should not signal debounce if the second invocation is late enough`() {
+        fixture.currentTimeMs = 1000
+        val debouncer = fixture.getDebouncer(3000)
+        assertFalse(debouncer.checkForDebounce())
+
+        fixture.currentTimeMs = 3001
```
Comment (stefanosiano) [inline sentry-android-core/src/test/java/io/sentry/android/core/internal/util/DebouncerTest.kt:59]: this test is like the previous one
guess you meant this?

```
fixture.currentTimeMs = 4001
assertFalse(debouncer.checkForDebounce())
```

File: sentry-android-core/src/test/java/io/sentry/android/core/ViewHierarchyEventProcessorTest.kt [hunk 3]
```diff
@@ -315,6 +315,72 @@ class ViewHierarchyEventProcessorTest {
...
+        val (_, hint) = fixture.process(
+            true,
+            SentryEvent().apply {
+                exceptions = listOf(SentryException())
+            }
+        )
+
+        assertNotNull(hint.viewHierarchy)
+    }
+
```
Comment (stefanosiano) [inline sentry-android-core/src/test/java/io/sentry/android/core/ViewHierarchyEventProcessorTest.kt:383]: could you add a last test `when capture callback returns true, debounce flag is overwritten`?
basically a test where the debounce flag is false and the callback returns true, so the view hierarchy should be attached anyway

File: sentry-android-core/src/test/java/io/sentry/android/core/ScreenshotEventProcessorTest.kt [hunk 4]
```diff
@@ -200,5 +202,81 @@ class ScreenshotEventProcessorTest {
...
+
+        val event = SentryEvent().apply {
+            exceptions = listOf(SentryException())
+        }
+        val hint = Hint()
+
+        processor.process(event, hint)
+        assertNotNull(hint.screenshot)
+    }
+
```
Comment (stefanosiano) [inline sentry-android-core/src/test/java/io/sentry/android/core/ScreenshotEventProcessorTest.kt:280]: could you add a last test `when capture callback returns true, debounce flag is overwritten`?
basically a test where the debounce flag is false and the callback returns true, so the view screenshot should be attached anyway

File: sentry-android-core/src/main/java/io/sentry/android/core/ViewHierarchyEventProcessor.java [hunk 5]
```diff
@@ -59,6 +66,16 @@ public ViewHierarchyEventProcessor(final @NotNull SentryAndroidOptions options)
       return event;
     }
 
+    final boolean shouldDebounce = debouncer.checkForDebounce();
```
Comment (stefanosiano) [inline sentry-android-core/src/main/java/io/sentry/android/core/ViewHierarchyEventProcessor.java:69]: [same comment](https://github.com/getsentry/sentry-java/pull/2773/files#r1219084571) here

File: sentry-android-core/src/test/java/io/sentry/android/core/ViewHierarchyEventProcessorTest.kt (hunk 3 above)
Comment (markushi) [reply sentry-android-core/src/test/java/io/sentry/android/core/ViewHierarchyEventProcessorTest.kt:383]: Oh yeah I missed that, thanks! This test case actually caught an edge case 🙇

**Review by stefanosiano — APPROVED** [review]

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java [hunk 6]
```diff
@@ -52,6 +59,19 @@ public ScreenshotEventProcessor(
       return event;
     }
 
+    // skip capturing in case of debouncing (=too many frequent capture requests)
+    // the BeforeCaptureCallback may overrules the debouncing decision
+    final boolean shouldDebounce = debouncer.checkForDebounce();
+    final @Nullable SentryAndroidOptions.BeforeCaptureCallback beforeCaptureCallback =
+        options.getBeforeScreenshotCaptureCallback();
+    if (beforeCaptureCallback != null) {
```
Comment (romtsn) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:67]: just checking if that's what we want - if the `beforeCaptureCallback` is set, we will *never* debounce (or rather - we completely move this responsibility to our users). Maybe it's worth to give more details in the javadoc at least? (or maybe docs, not sure).

**Review by romtsn — APPROVED** [review]

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java (hunk 6 above)
Comment (markushi) [reply sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:67]: Yes, exactly that's intentional - the user should have full control over this - let me update the docs then.

File: sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java [hunk 7]
```diff
@@ -52,6 +59,19 @@ public ScreenshotEventProcessor(
...
+      if (!beforeCaptureCallback.execute(event, hint, shouldDebounce)) {
```
Comment (krystofwoldrich) [inline sentry-android-core/src/main/java/io/sentry/android/core/ScreenshotEventProcessor.java:68]: In this case, the `shouldDebounce` works as a recommendation for the user and they can decide if they will respect it or not, right?
